### PR TITLE
feat: question rating and history backend

### PR DESCRIPTION
**Pull request by neojxuan**
1. update endpoint for questions to include ratings 
2. added endpoint to add a question attempt
3. added endpoint to complete/submit a question attempt (update the `AttemptedQuestions` with `code` and `completed_at`)

Schema Changes
1. added `code` column to `AttemptedQuestions`
4. made `code` and `completed_at` columns for AttemptedQuestions optional (if not completed_at is automatically filled to be the creation time which may be confusing)

### CODE REVIEW

File: backend/questions/src/questions/questionRouter.ts [hunk 1]
```diff
@@ -98,7 +134,12 @@ questionRouter.put(
     const updateData: QuestionUpdateData = {
       title: title || existingQuestion.title,
       body: body || existingQuestion.body,
-      complexity: complexity || existingQuestion.complexity
+      complexity: complexity || existingQuestion.complexity,
+      avgRating: ratings
+        ? (existingQuestion.ratings * existingQuestion.avgRating + ratings) /
+          (existingQuestion.ratings + 1)
+        : existingQuestion.avgRating,
+      ratings: ratings ? existingQuestion.ratings + 1 : existingQuestion.ratings
```
Comment (markusyeo) [inline backend/questions/src/questions/questionRouter.ts:150]: We need to wrap the truthy check here to check if it is 0 too, or we can just close 2 eyes here and reject 0 ratings xD

Comment (neojxuan) [reply backend/questions/src/questions/questionRouter.ts:150]: Hmmm shall we validate that rating is in [1, 5]?

File: backend/questions/src/questions/questionRouter.ts [hunk 2]
```diff
@@ -138,3 +179,50 @@ questionRouter.delete(
...
+    // Respond with a success message
+    res.status(201).json({
+      message: 'Question attempt created successfully',
+      questionAttempt: newAttempt
+    });
+  })
+);
+
+// Complete question attempt
+questionRouter.put(
```
Comment (markusyeo) [inline backend/questions/src/questions/questionRouter.ts:214]: nice yeah we'll need this

File: backend/questions/src/questions/questionRouter.ts (hunk 1 above)
Comment (markusyeo) [reply backend/questions/src/questions/questionRouter.ts:150]: Yeah we could, then I'll take note of the ui similarly, and do the validation on frontend as well

Comment (neojxuan) [reply backend/questions/src/questions/questionRouter.ts:150]: ok thanks! Hahaha I realise I still need to check if it's zero...Pushed the fix

**Comment by markusyeo** [comment]
LGTM

**Pull request merged**